core/src/mast/debuginfo/node_decorator_storage.rs

```diff
@@ -0,0 +1,212 @@
+        self.node_indptr_for_after[MastNodeId::new_unchecked((node_id.to_usize() + 1) as u32)] =
+            after_end;
+    }
+
+    /// Gets the before_enter decorators for a given node.
+    pub fn get_before_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        let node_idx = node_id.to_usize();
+
+        // Check if we have pointers for this node
+        if node_idx + 1 >= self.node_indptr_for_before.len() {
+            return &[];
+        }
+
+        let start = self.node_indptr_for_before[node_id];
+        let end = self.node_indptr_for_before[MastNodeId::new_unchecked((node_idx + 1) as u32)];
+
+        if start > end || end > self.before_enter_decorators.len() {
+            return &[];
+        }
+
+        &self.before_enter_decorators[start..end]
+    }
+
+    /// Gets the after_exit decorators for a given node.
+    pub fn get_after_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        let node_idx = node_id.to_usize();
+
+        // Check if we have pointers for this node
+        if node_idx + 1 >= self.node_indptr_for_after.len() {
+            return &[];
+        }
+
+        let start = self.node_indptr_for_after[node_id];
+        let end = self.node_indptr_for_after[MastNodeId::new_unchecked((node_idx + 1) as u32)];
+
+        if start > end || end > self.after_exit_decorators.len() {
+            return &[];
+        }
+
+        &self.after_exit_decorators[start..end]
+    }
+
+    /// Finalizes the storage by ensuring sentinel pointers are properly set.
+    /// This should be called after all nodes have been added.
+    pub fn finalize(&mut self) {
+        // Ensure sentinel pointers exist for all nodes
+        let max_len = self.node_indptr_for_before.len().max(self.node_indptr_for_after.len());
+
+        // Add final sentinel pointers if needed
+        if self.node_indptr_for_before.len() == max_len {
+            let _ = self.node_indptr_for_before.push(self.before_enter_decorators.len());
+        }
+        if self.node_indptr_for_after.len() == max_len {
+            let _ = self.node_indptr_for_after.push(self.after_exit_decorators.len());
+        }
+    }
+
+    /// Removes all decorators for a given node.
+    ///
+    /// # Arguments
+    /// * `_node_id` - The ID of the node to remove decorators for
+    ///
+    /// Note: This operation is not supported in the CSR structure as it would require
+    /// shifting all subsequent data. This method is a no-op.
+    pub fn remove_decorators(&mut self, _node_id: MastNodeId) {
+        // For CSR structure, removing individual node decorators is not supported
+        // as it would require shifting all subsequent data. This is a no-op.
+    }
+
+    /// Clears all decorators and mappings.
+    pub fn clear(&mut self) {
+        self.before_enter_decorators.clear();
+        self.after_exit_decorators.clear();
+        self.node_indptr_for_before = IndexVec::new();
+        self.node_indptr_for_after = IndexVec::new();
+    }
```

Comment:
It's not immediately clear to me the difference between `clear()` and `remove_decorators()` - since "remove decorators" is a no-op, but "clear all decorators" isnt.

This might just be a documentation/naming issue.